### PR TITLE
Add (F)GW barycenters

### PR DESCRIPTION
**Pull request by michalk8**
Also adds continuous barycenters to docs.

TODOs:
- [ ] verify that GW implementation is correct
- [ ] verify that FGW implementation is correct
- [ ] add tests
- [x] polish docs (esp. add references).
- [x] add utility function to pad cost matrices?
- ~~[ ] add example usage/tutorial?~~ in a separate PR

### CODE REVIEW

File: docs/core.rst [hunk 1]
```diff
@@ -45,13 +46,18 @@ Barycenters (Entropic and LR)
     :toctree: _autosummary
 
     discrete_barycenter.discrete_barycenter
+    continuous_barycenter.WassersteinBarycenter
+    continuous_barycenter.BarycenterState
+    gw_barycenter.GromovWassersteinBarycenter
```
Comment (marcocuturi) [inline docs/core.rst:51]: maybe worth adding a few refs in docs on this, probably our GW barycenter paper + the FGW work by Vayer Titouan, Chapel Laetitia, Flamary Rémi, Tavenard Romain and Courty Nicolas “Optimal Transport for structured data with application on graphs” International Conference on Machine Learning (ICML). 2019.

File: docs/core.rst [hunk 2]
```diff
@@ -63,6 +69,13 @@ Neural Potentials
     neuraldual.NeuralDualSolver
     neuraldual.NeuralDual
 
+Utilities
```
Comment (marcocuturi) [inline docs/core.rst:72]: maybe be more specific? Padding Utilities?

File: ott/core/bar_problems.py [hunk 3]
```diff
@@ -116,30 +121,37 @@ def add_slice_for_debiased(
 
   @property
   def flattened_y(self) -> Optional[jnp.ndarray]:
+    """Array of shape ``[num_measures * N, D]``."""
```
Comment (marcocuturi) [inline ott/core/bar_problems.py:124]: technically this is only valid if all measures have the same size. Maybe *(N_1 + N_2 + ... + N_num_measures) ?

File: ott/core/bar_problems.py [hunk 4]
```diff
@@ -116,30 +121,37 @@ def add_slice_for_debiased(
...
     if self._y is not None and self._y.ndim == 3:
       return self._y.reshape((-1, self._y.shape[-1]))
     else:
       return self._y
 
   @property
   def flattened_b(self) -> Optional[jnp.ndarray]:
+    """Array of shape ``[num_measures * N,]``."""
```
Comment (marcocuturi) [inline ott/core/bar_problems.py:132]: same comment

File: ott/core/bar_problems.py [hunk 5]
```diff
@@ -152,6 +164,7 @@ def max_measure_size(self) -> int:
 
   @property
   def num_segments(self) -> int:
+    """Number of measures."""
     if self._y is None:
```
Comment (marcocuturi) [inline ott/core/bar_problems.py:168]: I think you defined this to be no longer optional?

Comment (marcocuturi) [reply ott/core/bar_problems.py:168]: or maybe this is just there to help with the jit "dry run" compilation?

File: ott/core/bar_problems.py [hunk 6]
```diff
@@ -177,3 +191,171 @@ def weights(self) -> jnp.ndarray:
...
+    kwargs: Keyword arguments for
+      :class:`ott.core.bar_problems.BarycenterProblem`.
+  """
+
+  def __init__(
+      self,
+      y: Optional[jnp.ndarray] = None,
+      b: Optional[jnp.ndarray] = None,
+      weights: Optional[jnp.ndarray] = None,
+      cost: Optional[jnp.ndarray] = None,
```
Comment (marcocuturi) [inline ott/core/bar_problems.py:231]: i think it might be more readable that cost comes right after y. maybe use costs. ys is ugly and that's why we stick to y, but cost in singular is a bit confusing.

File: ott/core/bar_problems.py [hunk 7]
```diff
@@ -177,3 +191,171 @@ def weights(self) -> jnp.ndarray:
...
+
+    if self._loss_name == 'kl':
+      barycenter = jnp.exp(barycenter)
+    return barycenter
+
+  def update_features(self, transports: jnp.ndarray,
+                      a: jnp.ndarray) -> Optional[jnp.ndarray]:
+    """Update the barycenter features. Only used in the fused cased.
+
+    Only implemented for :class:`~ott.geometry.costs.Euclidean` cost.
```
Comment (marcocuturi) [inline ott/core/bar_problems.py:295]: and power =2 ?

File: ott/core/gw_barycenter.py [hunk 8]
```diff
@@ -0,0 +1,323 @@
...
+    a: Weights of the barycenter of shape ``[B,]``.
+    errors: Array of shape
+      ``[max_iter, num_measures, quad_max_iter, lin_outer_iter]`` containing
+      the GW errors at each iteration.
+    costs: Array of shape ``[max_iter,]`` containing the cost at each iteration.
+    gw_convergence: Array of shape ``[max_iter,]`` containing the convergence
+      of all GW problems at each iteration.
+  """
+  x: Optional[jnp.ndarray] = None
+  c: Optional[jnp.ndarray] = None
```
Comment (marcocuturi) [inline ott/core/gw_barycenter.py:35]: i feels c should come first, and maybe call it cost to be clear that this is not a vector or a point cloud

File: ott/core/gw_barycenter.py [hunk 9]
```diff
@@ -0,0 +1,323 @@
...
+from ott.geometry import geometry, pointcloud
+
+__all__ = ["GWBarycenterState", "GromovWassersteinBarycenter"]
+
+
+class GWBarycenterState(NamedTuple):
+  """Holds the state of the :class:`ott.core.bar_problems.GWBarycenterProblem`.
+
+  Args:
+    x: Barycenter features of shape ``[B, D_f]``. Only used in the fused case.
```
Comment (marcocuturi) [inline ott/core/gw_barycenter.py:24]: here just a comment on the fact that B sounds like batch, whereas I feel k could be a bit more unified (this is what is used in continuous_barycenter)

File: ott/core/gw_barycenter.py [hunk 10]
```diff
@@ -0,0 +1,323 @@
...
+  def __call__(self, problem: GWBarycenterProblem, **kwargs: Any):
+    bar_fn = jax.jit(iterations, static_argnums=1) if self.jit else iterations
+    state = self.init_state(problem, **kwargs)
+    state = bar_fn(solver=self, problem=problem, init_state=state)
+    return self.output_from_state(state)
+
+  def init_state(
+      self,
+      problem: GWBarycenterProblem,
+      bar_init: Union[int, jnp.ndarray, Tuple[jnp.ndarray, jnp.ndarray]],
```
Comment (marcocuturi) [inline ott/core/gw_barycenter.py:105]: this feels like it hides a bit more than just an init, but an important hyperparameter (size of the barycenter)

Comment (marcocuturi) [reply ott/core/gw_barycenter.py:105]: maybe renaming this to keep the API you propose, or use something more like what is used for continuous_barycenter?

File: ott/core/gw_barycenter.py [hunk 11]
```diff
@@ -0,0 +1,323 @@
...
+    solver: Linear OT solver.
+    key: Random key.
+    a: Source marginals (e.g., for barycenter) of shape ``[B,]``.
+    b: Target marginals of shape ``[N,]``.
+    epsilon: Entropy regularization.
+
+  Returns:
+    Transport map of shape ``[B, N]``.
+  """
+  cost = jax.random.uniform(key, shape=(len(a), len(b)), minval=0, maxval=1)
```
Comment (marcocuturi) [inline ott/core/gw_barycenter.py:287]: intuitively, i guess it might make sense to initialize with something that looks indeed like a cost matrix, at least symmetric. maybe a random pointcloud in arbitrary dimensions (e.g. 2d?)

**Review by marcocuturi — COMMENTED** [review]
Thanks a lot Michal!!! a few remarks, will dive deeper upon your next commits

File: docs/core.rst (hunk 2 above)
Comment (michalk8) [reply docs/core.rst]: Done.

File: ott/core/bar_problems.py (hunk 5 above)
Comment (michalk8) [reply ott/core/bar_problems.py]: > I think you defined this to be no longer optional?

You're right. I've refactored the code after consulting with @ersisimou so that optionally segmentation happens in `__init__`. Tests locally passed, so hope it didn't break anything else.

File: ott/core/bar_problems.py (hunk 6 above)
Comment (michalk8) [reply ott/core/bar_problems.py]: Agreed, changed `cost` -> `costs.`

File: ott/core/bar_problems.py (hunk 7 above)
Comment (michalk8) [reply ott/core/bar_problems.py:253]: Done.

File: ott/core/gw_barycenter.py (hunk 8 above)
Comment (michalk8) [reply ott/core/gw_barycenter.py:35]: Renamed to `cost`.

File: ott/core/gw_barycenter.py (hunk 9 above)
Comment (michalk8) [reply ott/core/gw_barycenter.py:24]: Good point, will change it to `k`. Generally do you prefer for the dim names to be lower case or upper case?

File: ott/core/gw_barycenter.py (hunk 11 above)
Comment (michalk8) [reply ott/core/gw_barycenter.py]: Nice catch, was following the original impl. for consistency. Now sample a random PC in 2D.

File: docs/core.rst (hunk 1 above)
Comment (michalk8) [reply docs/core.rst:51]: I've introduced bibtex files to the documentation and added those references. In the future, would convert all other references to `.bib` files.

**Comment by michalk8** [comment]
Could we enable the doc build on PR? Thought it was already working. Here's the link: https://docs.readthedocs.io/en/stable/pull-requests.html#configuration

**Review by marcocuturi — APPROVED** [review]
Thanks a lot for this great PR Michal!

File: ott/core/segment.py [hunk 12]
```diff
@@ -88,3 +88,46 @@ def segment_point_cloud(
   segmented_a = jnp.stack(segmented_a)
   segmented_x = jnp.stack(segmented_x)
   return segmented_x, segmented_a, num_segments
+
+
+def pad_along_axis(
```
[the remark this answers is not in the document]
Comment (michalk8) [reply ott/core/segment.py:93]: Atm it isn't being used anywhere, intended for tutorials/generic utility to replace `segment_point_cloud`. Can remove for now.

[the remark this answers is not in the document]
Comment (marcocuturi) [reply ott/core/segment.py:105]: ok, no problem, as you wish. As a general rule I think it's better not to add features that are not tested or used elsewhere, but happy to make an exception for this one given it's quite simple.
